drivers/storage/portworx/component/csi.go

```diff
@@ -989,15 +1036,58 @@ func createCSINodeInfoCRD() error {
 	return apiextensionsops.Instance().ValidateCRDV1beta1(resource, 1*time.Minute, 5*time.Second)
 }
 
+func (c *csi) createCSISnapshotCRDs(k8sVersion *version.Version) error {
+	// Detect which CRD to use
+	var snapshotCRDVersion string
+	switch {
+	case k8sVersion.GreaterThanOrEqual(k8sutil.K8sVer1_17) && k8sVersion.LessThan(k8sutil.K8sVer1_20):
+		snapshotCRDVersion = "v3"
+
+	case k8sVersion.GreaterThanOrEqual(k8sutil.K8sVer1_20):
+		snapshotCRDVersion = "v4"
+
+	default:
+		// CRD creation not supported for K8s < 1.17
+		return nil
+	}
+
+	// Register all CRDs for a given version
+	for _, crdFile := range []string{csiCRDSuffixVolumeSnapshot, csiCRDSuffixVolumeSnapshotContent, csiCRDSuffixVolumeSnapshotClass} {
+		crd := &apiextensionsv1.CustomResourceDefinition{}
+
+		// Parse CRD from file
+		if err := k8sutil.ParseObjectFromFile(
+			path.Join(pxutil.SpecsBaseDir(), fmt.Sprintf("%s-%s-%s", csiCRDPrefix, snapshotCRDVersion, crdFile)),
+			c.scheme,
+			crd,
+		); err != nil {
+			return err
+		}
+
+		// Create CRD
+		if err := k8sutil.CreateCRD(crd); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
 func (c *csi) getCSIConfiguration(
 	cluster *corev1.StorageCluster,
 	pxVersion *version.Version,
 ) *pxutil.CSIConfiguration {
 	deprecatedCSIDriverName := pxutil.UseDeprecatedCSIDriverName(cluster)
 	disableCSIAlpha := pxutil.DisableCSIAlpha(cluster)
 	kubeletPath := pxutil.KubeletPath(cluster)
+	if cluster.Spec.CSI == nil {
+		cluster.Spec.CSI = &corev1.CSISpec{
+			Enabled:                   false,
+			InstallSnapshotController: boolPtr(false),
+		}
+	}
```

Comment:
Why do we need this?

Reply:
There were some code paths breaking when the csi spec wasn't initialized. That code path also didn't fill the defaults.